### PR TITLE
Bug 1381749 - Add keywords to new-bug

### PR DESCRIPTION
**Pull request by SebastinSanty**

### CODE REVIEW

File: extensions/BugModal/lib/WebService.pm [hunk 1]
```diff
@@ -82,9 +82,10 @@ sub rest_resources {
     ]
 }
 
-sub products {
+sub initial_loader {
     my $user = Bugzilla->user;
-    return { products => _name($user->get_enterable_products) };
+    my @keywords = map { { name => $_->name } } grep { $_->is_active } Bugzilla::Keyword->get_all();
```
Comment (floatingatoll) [inline extensions/BugModal/lib/WebService.pm]: I'd be happier to see `my @products` here as well, so that +88 below isn't a mix of the two styles, but that's a tiny aesthetic nit and is no obstacle or anything.

File: extensions/BugModal/lib/WebService.pm [hunk 2]
```diff
@@ -30,9 +30,9 @@ sub rest_resources {
     return [
         # return all the products accessible by the user.
         # required by new-bug
-        qr{^/bug_modal/products}, {
+        qr{^/bug_modal/initial_loader}, {
```
Comment (floatingatoll) [inline extensions/BugModal/lib/WebService.pm]: Is this a safe change to make?

File: extensions/BugModal/lib/WebService.pm [hunk 3]
```diff
@@ -104,6 +105,7 @@ sub product_info {
         }
     } @{ $product->components };
     my @versions = map { { name => $_->name } } grep { $_->is_active } @{ $product->versions };
+    my @keywords = grep { $_->is_active } Bugzilla::Keyword->get_all();
```
Comment (floatingatoll) [inline extensions/BugModal/lib/WebService.pm]: `@keywords` isn't used after this; please either include it in the return value or add a code comment documenting the action at a distance.

File: extensions/BugModal/web/new_bug.js [hunk 4]
```diff
@@ -1,28 +1,28 @@
+var initial = {}
 var comp_desc = {}
 
-function load_products(query, callback) {
+$(document).ready(function() {
     bugzilla_ajax(
             {
-                url: 'rest/bug_modal/products'
+                url: 'rest/bug_modal/initial_loader'
```
Comment (floatingatoll) [inline extensions/BugModal/web/new_bug.js]: 'init' would also work well here, though it may be too abrupt/brief for practical use.

File: extensions/BugModal/web/new_bug.js [hunk 5]
```diff
@@ -1,28 +1,28 @@
...
             },
             function(data) {
-                callback(data.products);
+                initial = data
             },
             function() {
-                callback();
+                alert("Network issues. Please refresh the page and try again");
```
Comment (floatingatoll) [inline extensions/BugModal/web/new_bug.js]: How is this error message localized? Will any entered, unsubmitted, data be lost? Is it our responsibility to warn of that?

File: extensions/BugModal/web/new_bug.js [hunk 6]
```diff
@@ -38,17 +38,31 @@ $(document).ready(function() {
...
+        create: false,
+        load: function(query, callback) {
+            callback(initial.keywords);
+        }
+    });
+
     product_sel.on("change", function () {
         $('#product-throbber').show();
         $('#component').attr('disabled', true);
+        $('#version').attr('disabled', true);
```
Comment (floatingatoll) [inline extensions/BugModal/web/new_bug.js]: How does this change relate to the described commit message of 'add keywords' support?

File: extensions/BugModal/web/new_bug.js (hunk 5 above)
Comment (dylanwh) [inline extensions/BugModal/web/new_bug.js]: As a note, we do not do localization currently.

File: extensions/BugModal/web/new_bug.js (hunk 6 above)
Comment (dylanwh) [inline extensions/BugModal/web/new_bug.js]: Probably should remove this

File: extensions/BugModal/lib/WebService.pm [hunk 7]
```diff
@@ -30,9 +30,9 @@ sub rest_resources {
     return [
         # return all the products accessible by the user.
         # required by new-bug
-        qr{^/bug_modal/products}, {
+        qr{^/bug_modal/initial_loader}, {
             GET => {
-                method => 'products'
+                method => 'initial_loader'
```
Comment (dylanwh) [inline extensions/BugModal/lib/WebService.pm]: initial_field_values

File: extensions/BugModal/lib/WebService.pm (hunk 3 above)
Comment (dylanwh) [inline extensions/BugModal/lib/WebService.pm]: Probably a mistake, right?

File: extensions/BugModal/web/new_bug.js (hunk 6 above)
Comment (dylanwh) [inline extensions/BugModal/web/new_bug.js]: remove this unrelated change

File: extensions/BugModal/lib/WebService.pm (hunk 7 above)
Comment (SebastinSanty) [inline extensions/BugModal/lib/WebService.pm]: Yeah, sounds good. I forgot to change it.

File: extensions/BugModal/lib/WebService.pm (hunk 2 above)
Comment (SebastinSanty) [inline extensions/BugModal/lib/WebService.pm]: Yes, it is just used by new-bug, and the the data it provides is more than just products, hence the name change.

File: extensions/BugModal/lib/WebService.pm (hunk 3 above)
Comment (SebastinSanty) [inline extensions/BugModal/lib/WebService.pm]: Yes, it is a mistake.

File: extensions/BugModal/web/new_bug.js (hunk 6 above)
Comment (SebastinSanty) [inline extensions/BugModal/web/new_bug.js]: Yes, again a mistake. I'll change it. Anyways it doesn't make any UI change as selectize doesn't seem to heed to that instruction.

File: extensions/BugModal/web/new_bug.js (hunk 5 above)
Comment (SebastinSanty) [inline extensions/BugModal/web/new_bug.js]: No data will be lost. This is just initial data coming in.

**Comment by dylanwh** [comment]
in the POST processing part, keywords will be an array:
```perl
my @keywords = $cgi->param('keywords');
```

You can then pass `\@keywords` to `Bugzilla::Bug->new`

**Comment by dylanwh** [comment]
some changes in https://github.com/SebastinSanty/bmo/pull/5

**Comment by dylanwh** [comment]
@SebastinSanty can you handle the merge conflict? I think it's from the perltidy change.
